File: examples/Delphi/CustOrd2.pas
{$J+} {Writable constants}

unit CustOrd2;

interface

uses
  SysUtils, WinTypes, WinProcs, Messages, Classes, Graphics, Controls, Forms,
  Dialogs, OvcTCmmn, OvcTable, OvcPb, OvcPf, OvcTCPic, OvcBase, OvcEf, OvcSf,
  OvcTCSim, OvcTCBEF, OvcTCStr, OvcTCEdt, DBTables, DB, OvcData,
  OvcCache, {this is a bonus component. It should be in \orpheus\bonus}
  {and must be dearchived and installed prior to compiling this example}
  ExtCtrls, StdCtrls, Buttons, OvcTCHdr, OvcTCell, OvcDate;

const
  {number of records to cache}
  MaxCacheItems = 20;

type
  {definition of one record for order table}
  POrderData = ^TOrderData;
  TOrderData = record
    OrderNo       : Double; {*index}
    SaleDate      : TOvcDate;
    ShipDate      : TOvcDate;
    EmpNo         : LongInt;
    ShipToContact : string[20];
    ShipToAddr1   : string[30];
    ShipToAddr2   : string[30];
    ShipToCity    : string[15];
    ShipToState   : string[20];
    ShipToZip     : string[10];
    ShipToCountry : string[20];
    ShipToPhone   : string[15];
    ShipVIA       : string[7];
    PO            : string[15];
    Terms         : string[6];
    PaymentMethod : string[7];
    ItemsTotal    : Double;
    TaxRate       : Double;
    Freight       : Double;
    AmountPaid    : Double;
  end;

type
  TfrmOrders = class(TForm)
    dbOrderQuery: TQuery;
    dbOrderQueryOrderNo: TFloatField;
    dbOrderQuerySaleDate: TDateTimeField;
    dbOrderQueryShipDate: TDateTimeField;
    dbOrderQueryEmpNo: TIntegerField;
    dbOrderQueryShipToContact: TStringField;
    dbOrderQueryShipToAddr1: TStringField;
    dbOrderQueryShipToAddr2: TStringField;
    dbOrderQueryShipToCity: TStringField;
    dbOrderQueryShipToState: TStringField;
    dbOrderQueryShipToZip: TStringField;
    dbOrderQueryShipToCountry: TStringField;
    dbOrderQueryShipToPhone: TStringField;
    dbOrderQueryShipVIA: TStringField;
    dbOrderQueryPO: TStringField;
    dbOrderQueryTerms: TStringField;
    dbOrderQueryPaymentMethod: TStringField;
    dbOrderQueryItemsTotal: TCurrencyField;
    dbOrderQueryTaxRate: TFloatField;
    dbOrderQueryFreight: TCurrencyField;
    dbOrderQueryAmountPaid: TCurrencyField;
    sfOrderNo: TOvcTCSimpleField;
    pfSaleDate: TOvcTCPictureField;
    pfShipDate: TOvcTCPictureField;
    sfEmpNo: TOvcTCSimpleField;
    sfShipToContact: TOvcTCSimpleField;
    sfShipToAddr1: TOvcTCSimpleField;
    sfShipToAddr2: TOvcTCSimpleField;
    sfShipToCity: TOvcTCSimpleField;
    sfShipToState: TOvcTCSimpleField;
    pfShipToZip: TOvcTCPictureField;
    sfShipToCountry: TOvcTCSimpleField;
    pfShipToPhone: TOvcTCPictureField;
    sfShipToVia: TOvcTCSimpleField;
    sfPO: TOvcTCSimpleField;
    sfTerms: TOvcTCSimpleField;
    sfPaymentMethod: TOvcTCSimpleField;
    pfItemsTotal: TOvcTCPictureField;
    pfTaxRate2: TOvcTCPictureField;
    pfFreight: TOvcTCPictureField;
    pfAmountPaid: TOvcTCPictureField;
    tblOrder: TOvcTable;
    Panel2: TPanel;
    OvcController1: TOvcController;
    BitBtn1: TBitBtn;
    RowLable: TOvcTCRowHead;
    dbOrderHeaderCell: TOvcTCColHead;
    procedure tblOrderGetCellData(Sender: TObject; RowNum: Longint;
      ColNum: Integer; var Data: Pointer; Purpose : TOvcCellDataPurpose);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure tblOrderBeginEdit(Sender: TObject; RowNum: Longint;
      ColNum: Integer; var AllowIt: Boolean);
    procedure tblOrderActiveCellChanged(Sender: TObject; RowNum: Longint;
      ColNum: Integer);
    procedure tblOrderLockedCellClick(Sender: TObject; RowNum: Longint;
      ColNum: Integer);
    procedure BitBtn1Click(Sender: TObject);

  private
    CellData   : string;
    OrderCache : TOvcCache;

  public
    procedure GetOrderRecord(RowNum : LongInt; var P : Pointer; var Size : Word);
    procedure PositionDatabase(RowNum : LongInt);
    procedure RefreshRecord(RecNum : LongInt);
    procedure UpdateOrderTable(CustNo : Double);
  end;

var
  frmOrders: TfrmOrders;

implementation

{$R *.DFM}

procedure TfrmOrders.PositionDatabase(RowNum : LongInt);
const
  LastRecordRequest : LongInt = -1;
begin
  {position db cursor to requested record}
  if LastRecordRequest = -1 then begin
    dbOrderQuery.First;
    dbOrderQuery.MoveBy(RowNum-1);
  end else if RowNum = 1 then
    dbOrderQuery.First
  else if RowNum = dbOrderQuery.RecordCount then
    dbOrderQuery.Last
  else
    dbOrderQuery.MoveBy(RowNum-LastRecordRequest);
  LastRecordRequest := RowNum;
end;

procedure TfrmOrders.GetOrderRecord(RowNum : LongInt; var P : Pointer; var Size : Word);
begin
  {position db cursor to requested record}
  PositionDatabase(RowNum);
  {return size of data}
  Size := SizeOf(TOrderData);
  {allocate buffer for data}
  GetMem(P, Size);
  {read the db record}
  with POrderData(P)^ do begin
    OrderNo       := dbOrderQueryOrderNo.AsFloat;
    SaleDate      := DateTimeToStDate(dbOrderQuerySaleDate.AsDateTime);
    ShipDate      := DateTimeToStDate(dbOrderQueryShipDate.AsDateTime);
    EmpNo         := dbOrderQueryEmpNo.AsInteger;
    ShipToContact := ShortString(dbOrderQueryShipToContact.AsString);
    ShipToAddr1   := ShortString(dbOrderQueryShipToAddr1.AsString);
    ShipToAddr2   := ShortString(dbOrderQueryShipToAddr2.AsString);
    ShipToCity    := ShortString(dbOrderQueryShipToCity.AsString);
    ShipToState   := ShortString(dbOrderQueryShipToState.AsString);
    ShipToZip     := ShortString(dbOrderQueryShipToZip.AsString);
    ShipToCountry := ShortString(dbOrderQueryShipToCountry.AsString);
    ShipToPhone   := ShortString(dbOrderQueryShipToPhone.AsString);
    ShipVIA       := ShortString(dbOrderQueryShipVIA.AsString);
    PO            := ShortString(dbOrderQueryPO.AsString);
    Terms         := ShortString(dbOrderQueryTerms.AsString);
    PaymentMethod := ShortString(dbOrderQueryPaymentMethod.AsString);
    ItemsTotal    := dbOrderQueryItemsTotal.AsFloat;
    TaxRate       := dbOrderQueryTaxRate.AsFloat;
    Freight       := dbOrderQueryFreight.AsFloat;
    AmountPaid    := dbOrderQueryAmountPaid.AsFloat;
  end;
end;

procedure TfrmOrders.tblOrderGetCellData(Sender: TObject; RowNum: Longint;
  ColNum: Integer; var Data: Pointer; Purpose : TOvcCellDataPurpose);
var
  P : POrderData;
begin
  CellData := '?';
  Data := @CellData;
  if RowNum = 0 then begin
    case ColNum of
      0 : CellData := '';
      1 : CellData := dbOrderQueryOrderNo.DisplayLabel;
      2 : CellData := dbOrderQuerySaleDate.DisplayLabel;
      3 : CellData := dbOrderQueryShipDate.DisplayLabel;
      4 : CellData := dbOrderQueryEmpNo.DisplayLabel;
      5 : CellData := dbOrderQueryShipToContact.DisplayLabel;
      6 : CellData := dbOrderQueryShipToAddr1.DisplayLabel;
      7 : CellData := dbOrderQueryShipToAddr2.DisplayLabel;
      8 : CellData := dbOrderQueryShipToCity.DisplayLabel;
      9 : CellData := dbOrderQueryShipToState.DisplayLabel;
     10 : CellData := dbOrderQueryShipToZip.DisplayLabel;
     11 : CellData := dbOrderQueryShipToCountry.DisplayLabel;
     12 : CellData := dbOrderQueryShipToPhone.DisplayLabel;
     13 : CellData := dbOrderQueryShipVIA.DisplayLabel;
     14 : CellData := dbOrderQueryPO.DisplayLabel;
     15 : CellData := dbOrderQueryTerms.DisplayLabel;
     16 : CellData := dbOrderQueryPaymentMethod.DisplayLabel;
     17 : CellData := dbOrderQueryItemsTotal.DisplayLabel;
     18 : CellData := dbOrderQueryTaxRate.DisplayLabel;
     19 : CellData := dbOrderQueryFreight.DisplayLabel;
     20 : CellData := dbOrderQueryAmountPaid.DisplayLabel;
    end;
  end else begin
    P  := OrderCache.Items[RowNum];
    if Assigned(P) then begin
      case ColNum of
        0 : {};
        1 : Data := @P^.OrderNo;
        2 : Data := @P^.SaleDate;
        3 : Data := @P^.ShipDate;
        4 : Data := @P^.EmpNo;
        5 : Data := @P^.ShipToContact;
        6 : Data := @P^.ShipToAddr1;
        7 : Data := @P^.ShipToAddr2;
        8 : Data := @P^.ShipToCity;
        9 : Data := @P^.ShipToState;
       10 : Data := @P^.ShipToZip;
       11 : Data := @P^.ShipToCountry;
       12 : Data := @P^.ShipToPhone;
       13 : Data := @P^.ShipVIA;
       14 : Data := @P^.PO;
       15 : Data := @P^.Terms;
       16 : Data := @P^.PaymentMethod;
       17 : Data := @P^.ItemsTotal;
       18 : Data := @P^.TaxRate;
       19 : Data := @P^.Freight;
       20 : Data := @P^.AmountPaid;
      end
    end;
  end;
end;

procedure TfrmOrders.FormCreate(Sender: TObject);
begin
  OrderCache := TOvcCache.Create(Self);
  OrderCache.MaxCacheItems := MaxCacheItems;
  OrderCache.OnGetItem := GetOrderRecord;
end;

procedure TfrmOrders.FormDestroy(Sender: TObject);
begin
  dbOrderQuery.Close;
  OrderCache.Free;
end;

procedure TfrmOrders.RefreshRecord(RecNum : LongInt);
begin
  if Assigned(OrderCache) then begin                                   {!!.13}
    OrderCache.Update(RecNum);
    tblOrder.AllowRedraw := False;
    tblOrder.InvalidateRow(RecNum);
    tblOrder.AllowRedraw := True;
  end;                                                                 {!!.13}
end;

procedure TfrmOrders.UpdateOrderTable(CustNo : Double);
begin
  dbOrderQuery.Close;
  dbOrderQuery.Params[0].AsFloat := CustNo;
  try
    dbOrderQuery.RequestLive := False;
    dbOrderQuery.Open;
  except
    on EDatabaseError do begin
      tblOrder.RowLimit := 1;
      raise;
    end;
  end;
  OrderCache.Refresh;
  tblOrder.RowLimit := dbOrderQuery.RecordCount + 1;
  tblOrder.Refresh;
end;

procedure TfrmOrders.tblOrderBeginEdit(Sender: TObject; RowNum: Longint;
  ColNum: Integer; var AllowIt: Boolean);
begin
  {don't allow modifications to the order query}
  AllowIt := False;
end;

procedure TfrmOrders.tblOrderActiveCellChanged(Sender: TObject;
  RowNum: Longint; ColNum: Integer);
const
  LastRow : LongInt = -1;
begin
  {refresh row}
  if LastRow <> RowNum then begin
    LastRow := RowNum;
    RefreshRecord(RowNum);
  end;
end;

procedure TfrmOrders.tblOrderLockedCellClick(Sender: TObject;
  RowNum: Longint; ColNum: Integer);
begin
  if RowNum = 0 then
    tblOrder.ActiveCol := ColNum;
  if ColNum = 0 then
    tblOrder.ActiveRow := RowNum;
end;

procedure TfrmOrders.BitBtn1Click(Sender: TObject);
begin
  Close;
end;

end.

File: examples/CBuildr6/custord2.cpp
//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include "CustOrd2.h"
#include "CustOrd1.h"
//---------------------------------------------------------------------------
#pragma link "OvcTable"
#pragma link "OvcTCmmn"
#pragma link "OvcBase"
#pragma link "OvcTCSim"
#pragma link "OvcTCBEF"
#pragma link "OvcTCStr"
#pragma link "OvcTCell"
#pragma link "OvcTCPic"
#pragma link "OvcTCHdr"
#pragma resource "*.dfm"
TfrmOrders *frmOrders;

extern int MaxCacheItems;

//---------------------------------------------------------------------------
__fastcall TfrmOrders::TfrmOrders(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TfrmOrders::GetOrderRecord(int RowNum, void* &P, Word& Size)
{
  // position db cursor to requested record
  PositionDatabase(RowNum);
  // return size of data
  Size = sizeof(TOrderData);
  // allocate buffer for data
  P = new TOrderData;
  // Cast the void pointer to a TOrderData pointer. I used a pointer
  // here but I could have used a reference instead. Using a pointer
  // we have to use the -> operator rather than the dot operator.
  // To see the difference look at the corresponding method in the
  // CUSTORD1.CPP file. There I used a refernce rather than a pointer.
  TOrderData* data = reinterpret_cast<TOrderData*>(P);
  // read the db record
  data->OrderNo       = dbOrderQueryOrderNo->AsFloat;
  data->SaleDate      = DateTimeToStDate(dbOrderQuerySaleDate->AsDateTime);
  data->ShipDate      = DateTimeToStDate(dbOrderQueryShipDate->AsDateTime);
  data->EmpNo         = dbOrderQueryEmpNo->AsInteger;
  data->ShipToContact = dbOrderQueryShipToContact->AsString;
  data->ShipToAddr1   = dbOrderQueryShipToAddr1->AsString;
  data->ShipToAddr2   = dbOrderQueryShipToAddr2->AsString;
  data->ShipToCity    = dbOrderQueryShipToCity->AsString;
  data->ShipToState   = dbOrderQueryShipToState->AsString;
  data->ShipToZip     = dbOrderQueryShipToZip->AsString;
  data->ShipToCountry = dbOrderQueryShipToCountry->AsString;
  data->ShipToPhone   = dbOrderQueryShipToPhone->AsString;
  data->ShipVIA       = dbOrderQueryShipVIA->AsString;
  data->PO            = dbOrderQueryPO->AsString;
  data->Terms         = dbOrderQueryTerms->AsString;
  data->PaymentMethod = dbOrderQueryPaymentMethod->AsString;
  data->ItemsTotal    = dbOrderQueryItemsTotal->AsFloat;
  data->TaxRate       = dbOrderQueryTaxRate->AsFloat;
  data->Freight       = dbOrderQueryFreight->AsFloat;
  data->AmountPaid    = dbOrderQueryAmountPaid->AsFloat;
}

void TfrmOrders::PositionDatabase(int RowNum)
{
  static int LastRecordRequest = -1;
  // position db cursor to requested record
  if (LastRecordRequest == -1) {
    dbOrderQuery->First();
    dbOrderQuery->MoveBy(RowNum - 1);
  }
  else if (RowNum == 1)
    dbOrderQuery->First();
  else if (RowNum == dbOrderQuery->RecordCount)
    dbOrderQuery->Last();
  else
    dbOrderQuery->MoveBy(RowNum - LastRecordRequest);
  LastRecordRequest = RowNum;
}

void TfrmOrders::RefreshRecord(int RecNum)
{
  OrderCache->Update(RecNum);
  tblOrder->AllowRedraw = false;
  tblOrder->InvalidateRow(RecNum);
  tblOrder->AllowRedraw = true;
}

void TfrmOrders::UpdateOrderTable(double CustNo)
{
  dbOrderQuery->Close();
  dbOrderQuery->Params->Items[0]->AsFloat = CustNo;
  try {
    dbOrderQuery->RequestLive = false;
    dbOrderQuery->Open();
  }
  catch (EDatabaseError&) {
    tblOrder->RowLimit = 1;
    throw;
  }
  OrderCache->Refresh();
  tblOrder->RowLimit = dbOrderQuery->RecordCount + 1;
  tblOrder->Refresh();
}

void __fastcall TfrmOrders::tblOrderGetCellData(TObject *Sender, int RowNum,
  int ColNum, Pointer &Data, TOvcCellDataPurpose Purpose)
{
  TOrderData* P = 0;
  CellData = "?";
  Data = &CellData;
  if (RowNum == 0) {
    switch (ColNum) {
      case 0 : CellData = ""; break;
      case 1 : CellData = dbOrderQueryOrderNo->DisplayLabel; break;
      case 2 : CellData = dbOrderQuerySaleDate->DisplayLabel; break;
      case 3 : CellData = dbOrderQueryShipDate->DisplayLabel; break;
      case 4 : CellData = dbOrderQueryEmpNo->DisplayLabel; break;
      case 5 : CellData = dbOrderQueryShipToContact->DisplayLabel; break;
      case 6 : CellData = dbOrderQueryShipToAddr1->DisplayLabel; break;
      case 7 : CellData = dbOrderQueryShipToAddr2->DisplayLabel; break;
      case 8 : CellData = dbOrderQueryShipToCity->DisplayLabel; break;
      case 9 : CellData = dbOrderQueryShipToState->DisplayLabel; break;
      case 10 : CellData = dbOrderQueryShipToZip->DisplayLabel; break;
      case 11 : CellData = dbOrderQueryShipToCountry->DisplayLabel; break;
      case 12 : CellData = dbOrderQueryShipToPhone->DisplayLabel; break;
      case 13 : CellData = dbOrderQueryShipVIA->DisplayLabel; break;
      case 14 : CellData = dbOrderQueryPO->DisplayLabel; break;
      case 15 : CellData = dbOrderQueryTerms->DisplayLabel; break;
      case 16 : CellData = dbOrderQueryPaymentMethod->DisplayLabel; break;
      case 17 : CellData = dbOrderQueryItemsTotal->DisplayLabel; break;
      case 18 : CellData = dbOrderQueryTaxRate->DisplayLabel; break;
      case 19 : CellData = dbOrderQueryFreight->DisplayLabel; break;
      case 20 : CellData = dbOrderQueryAmountPaid->DisplayLabel; break;
    }
  }
  else {
    P = reinterpret_cast<TOrderData*>(OrderCache->Items[RowNum]);
    if (P) {
      switch (ColNum) {
        case 0 :
        case 1 : Data = &P->OrderNo; break;
        case 2 : Data = &P->SaleDate; break;
        case 3 : Data = &P->ShipDate; break;
        case 4 : Data = &P->EmpNo; break;
        case 5 : Data = &P->ShipToContact; break;
        case 6 : Data = &P->ShipToAddr1; break;
        case 7 : Data = &P->ShipToAddr2; break;
        case 8 : Data = &P->ShipToCity; break;
        case 9 : Data = &P->ShipToState; break;
        case 10 : Data = &P->ShipToZip;  break;
        case 11 : Data = &P->ShipToCountry; break;
        case 12 : Data = &P->ShipToPhone; break;
        case 13 : Data = &P->ShipVIA; break;
        case 14 : Data = &P->PO; break;
        case 15 : Data = &P->Terms; break;
        case 16 : Data = &P->PaymentMethod; break;
        case 17 : Data = &P->ItemsTotal; break;
        case 18 : Data = &P->TaxRate; break;
        case 19 : Data = &P->Freight; break;
        case 20 : Data = &P->AmountPaid;
      }
    }
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmOrders::FormCreate(TObject *Sender)
{
  OrderCache = new TOvcCache(this);
  OrderCache->MaxCacheItems = MaxCacheItems;
  OrderCache->OnGetItem = &GetOrderRecord;
}
//---------------------------------------------------------------------------
void __fastcall TfrmOrders::FormDestroy(TObject *Sender)
{
  dbOrderQuery->Close();
  delete OrderCache;
}
//---------------------------------------------------------------------------
void __fastcall TfrmOrders::tblOrderBeginEdit(TObject *Sender, int RowNum,
  int ColNum, bool &AllowIt)
{
  // don't allow modifications to the order query
  AllowIt = false;
}
//---------------------------------------------------------------------------
void __fastcall TfrmOrders::tblOrderActiveCellChanged(TObject *Sender,
  int RowNum, int ColNum)
{
  static int LastRow = -1;
  // refresh row
  if (LastRow != RowNum) {
    LastRow = RowNum;
    RefreshRecord(RowNum);
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmOrders::tblOrderLockedCellClick(TObject *Sender,
  int RowNum, int ColNum)
{
  if (RowNum == 0)
    tblOrder->ActiveCol = ColNum;
  if (ColNum == 0)
    tblOrder->ActiveRow = RowNum;
}
//---------------------------------------------------------------------------
void __fastcall TfrmOrders::BitBtn1Click(TObject *Sender)
{
  Close();    
}
//---------------------------------------------------------------------------

